sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala

```diff
@@ -92,271 +104,219 @@ private[lf] final class ValueTranslator(
-                      )
-                  }
-
-                if (config.enableUpgrade) {
-                  val oLabeledFlds =
-                    labeledRecordToMap(sourceElements)
-                      .fold(typeError, identity _)
-
-                  // correctFields: (correct only by label/position) gives the value and type, length == targetFieldsAndTypes
-                  //   filled with Nones when type is Optional
-                  // extraFields: Unknown additional fields with name and value
-                  val (correctFields, extraFields): (
-                      Seq[(Ref.Name, Value, Type)],
-                      ImmArray[(Option[Ref.Name], Value)],
-                  ) =
-                    oLabeledFlds match {
-                      // Not fully labelled (or reordering disabled), so order dependent
-                      // Additional fields should downgrade, missing fields should upgrade
-                      case None => {
-                        val correctFields = targetFieldsAndTypes.toSeq.zipWithIndex.map {
-                          case ((lbl, ty), i) =>
-                            val (mbLbl, v) =
-                              sourceElements.get(i).getOrElse(addMissingField(lbl, ty))
-                            mbLbl.foreach(lbl_ =>
-                              if (lbl_ != lbl)
-                                typeError(
-                                  s"Mismatching record field label '$lbl_' (expecting '$lbl') for record $tyCon"
-                                )
+          case (ContractIdF(_), ValueContractId(c)) =>
+            unsafeTranslateCid(c)
+          case (OptionalF(a), ValueOptional(mbValue)) =>
+            mbValue match {
+              case Some(v) =>
+                SValue.SOptional(Some(go(a, v, newNesting)))
+              case None =>
+                SValue.SValue.None
+            }
+          case (ListF(a), ValueList(ls)) =>
+            if (ls.isEmpty) {
+              SValue.SValue.EmptyList
+            } else {
+              SValue.SList(ls.toImmArray.toSeq.map(go(a, _, newNesting)).toImmArray.toFrontStack)
+            }
+          case (MapF(a, b), ValueGenMap(entries)) =>
+            if (entries.isEmpty) {
+              SValue.SValue.EmptyGenMap
+            } else {
+              SValue.SMap(
+                isTextMap = false,
+                entries = entries.toSeq.view.map { case (k, v) =>
+                  go(a, k, newNesting) -> go(b, v, newNesting)
+                },
+              )
+            }
+          case (TextMapF(a), ValueTextMap(entries)) =>
+            if (entries.isEmpty) {
+              SValue.SValue.EmptyGenMap
+            } else {
+              SValue.SMap(
+                isTextMap = true,
+                entries = entries.toImmArray.toSeq.view.map { case (k, v) =>
+                  SValue.SText(k) -> go(a, v, newNesting)
+                }.toList,
```

Comment:
Why `toImmArray` and then `toList`? I'm sure there's a good reason so it's more for my education than questioning your choice here. Happy to discuss offline.